stdlib/public/core/While.swift.gyb

```diff
@@ -0,0 +1,395 @@
+
+  public typealias Elements = LazyPrefixWhileSequence
+
+  public func makeIterator() -> LazyPrefixWhileIterator<Base.Iterator> {
+    return LazyPrefixWhileIterator(
+      _base: _base.makeIterator(), predicate: _predicate)
+  }
+
+  internal init(_base: Base, predicate: (Base.Iterator.Element) -> Bool) {
+    self._base = _base
+    self._predicate = predicate
+  }
+
+  internal var _base: Base
+  internal let _predicate: (Base.Iterator.Element) -> Bool
+}
+
+extension LazySequenceProtocol {
+  /// Returns a lazy sequence of the initial consecutive elements that satisfy
+  /// `predicate`.
+  ///
+  /// - Parameter predicate: A closure that takes an element of the sequence as
+  ///   its argument and returns `true` if the element should be included or
+  ///   `false` otherwise. Once `predicate` returns `false` it will not be
+  ///   called again.
+  public func prefix(
+    while predicate: (Elements.Iterator.Element) -> Bool
+  ) -> LazyPrefixWhileSequence<Self.Elements> {
+    return LazyPrefixWhileSequence(_base: self.elements, predicate: predicate)
+  }
+}
+
+//===--- Collections ------------------------------------------------------===//
+
+/// A position in a `LazyPrefixWhileCollection` or
+/// `LazyPrefixWhileBidirectionalCollection` instance.
+public struct LazyPrefixWhileIndex<Base : Collection> : Comparable {
+  /// The position corresponding to `self` in the underlying collection.
+  public let base: Base.Index
+}
+
+public func == <Base : Collection>(
+  lhs: LazyPrefixWhileIndex<Base>,
+  rhs: LazyPrefixWhileIndex<Base>
+) -> Bool {
+  return lhs.base == rhs.base
+}
+
+public func < <Base : Collection>(
+  lhs: LazyPrefixWhileIndex<Base>,
+  rhs: LazyPrefixWhileIndex<Base>
+) -> Bool {
+  return lhs.base < rhs.base
+}
+
+% for Traversal in ['Forward', 'Bidirectional']:
+%   Collection = collectionForTraversal(Traversal)
+%   Self = "LazyPrefixWhile" + Collection
+
+/// A lazy `${Collection}` wrapper that includes the initial consecutive
+/// elements of an underlying collection that satisfy a predicate.
+///
+/// - Note: The performance of accessing `endIndex`, `last`, any methods that
+///   depend on `endIndex`, or moving an index depends on how many elements
+///   satisfy the predicate at the start of the collection, and may not offer
+///   the usual performance given by the `Collection` protocol. Be aware,
+///   therefore, that general operations on `${Self}` instances may not have
+///   the documented complexity.
+public struct ${Self}<
+  Base : ${Collection}
+> : LazyCollectionProtocol, ${Collection} {
+
+  public typealias Index = LazyPrefixWhileIndex<Base>
+
+  public var startIndex: Index {
+    return LazyPrefixWhileIndex(base: _base.startIndex)
+  }
+
+  public var endIndex: Index {
+    var index = _base.startIndex
+    while index != _base.endIndex && _predicate(_base[index]) {
+      _base.formIndex(after: &index)
+    }
+    return LazyPrefixWhileIndex(base: index)
+  }
```

Comment:
I was going to ask a question about this very thing, since an O(n) `endIndex` does seem much worse than an O(n) `startIndex`. The main issue here is the requirement that `startIndex == endIndex` when the collection is empty.

Thinking about it now, though, we can check whether the collection is empty simply by testing the first element—if it fails the predicate, the collection is empty and we return `startIndex`, otherwise we return a specific end value without evaluating the rest of the collection.